File: backend/app/services/subtitle_generator.py
from pathlib import Path
from typing import List

from app.models.schemas import SubtitleEntry


class SubtitleGenerator:
    def generate_srt(
        self,
        subtitles: List[SubtitleEntry],
        output_path: Path,
        bilingual: bool = True,
    ) -> Path:
        srt_content = self._format_srt(subtitles, bilingual)
        output_path.write_text(srt_content, encoding="utf-8")
        return output_path

    def _format_srt(self, subtitles: List[SubtitleEntry], bilingual: bool) -> str:
        blocks = []
        for sub in subtitles:
            start = self._format_srt_time(sub.start_time)
            end = self._format_srt_time(sub.end_time)

            if bilingual:
                text = f"{sub.source_text}\n{sub.translated_text}"
            else:
                text = sub.translated_text

            block = f"{sub.index}\n{start} --> {end}\n{text}\n"
            blocks.append(block)

        return "\n".join(blocks)

    @staticmethod
    def _format_srt_time(time_str: str) -> str:
        if "," in time_str or ":" in time_str:
            return time_str

        try:
            seconds = float(time_str)
        except ValueError:
            return time_str

        hours = int(seconds // 3600)
        minutes = int((seconds % 3600) // 60)
        secs = int(seconds % 60)
        millis = int((seconds % 1) * 1000)
        return f"{hours:02d}:{minutes:02d}:{secs:02d},{millis:03d}"

    def parse_srt(self, srt_path: Path) -> List[SubtitleEntry]:
        content = srt_path.read_text(encoding="utf-8")
        blocks = content.strip().split("\n\n")
        entries = []

        for block in blocks:
            lines = block.strip().split("\n")
            if len(lines) < 3:
                continue

            index = int(lines[0])
            time_line = lines[1]
            start, end = time_line.split(" --> ")

            text_lines = lines[2:]
            if len(text_lines) >= 2:
                source_text = text_lines[0]
                translated_text = text_lines[1]
            else:
                source_text = ""
                translated_text = text_lines[0] if text_lines else ""

            entries.append(SubtitleEntry(
                index=index,
                start_time=start.strip(),
                end_time=end.strip(),
                source_text=source_text,
                translated_text=translated_text,
            ))

        return entries

File: backend/src/main/java/com/bili/translator/service/SubtitleGenerator.java
package com.bili.translator.service;

import com.bili.translator.model.SubtitleEntry;
import org.springframework.stereotype.Service;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

@Service
public class SubtitleGenerator {

    public void generateSrt(List<SubtitleEntry> subtitles, Path outputPath, boolean bilingual) throws Exception {
        String srtContent = formatSrt(subtitles, bilingual);
        byte[] bom = {(byte) 0xEF, (byte) 0xBB, (byte) 0xBF};
        byte[] contentBytes = srtContent.getBytes(StandardCharsets.UTF_8);
        byte[] withBom = new byte[bom.length + contentBytes.length];
        System.arraycopy(bom, 0, withBom, 0, bom.length);
        System.arraycopy(contentBytes, 0, withBom, bom.length, contentBytes.length);
        Files.write(outputPath, withBom);
    }

    private String formatSrt(List<SubtitleEntry> subtitles, boolean bilingual) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < subtitles.size(); i++) {
            SubtitleEntry sub = subtitles.get(i);
            String start = formatSrtTime(sub.getStartTime());
            String end = formatSrtTime(sub.getEndTime());

            String text;
            if (bilingual) {
                text = sub.getSourceText() + "\n" + sub.getTranslatedText();
            } else {
                text = sub.getTranslatedText();
            }

            sb.append(sub.getIndex()).append("\n");
            sb.append(start).append(" --> ").append(end).append("\n");
            sb.append(text).append("\n");
            if (i < subtitles.size() - 1) {
                sb.append("\n");
            }
        }
        return sb.toString();
    }

    private String formatSrtTime(String timeStr) {
        if (timeStr.contains(",") || timeStr.contains(":")) {
            return timeStr;
        }
        try {
            double seconds = Double.parseDouble(timeStr);
            return secondsToSrtTime(seconds);
        } catch (NumberFormatException e) {
            return timeStr;
        }
    }

    public static String secondsToSrtTime(double seconds) {
        int hours = (int) (seconds / 3600);
        int minutes = (int) ((seconds % 3600) / 60);
        int secs = (int) (seconds % 60);
        int millis = (int) ((seconds % 1) * 1000);
        return String.format("%02d:%02d:%02d,%03d", hours, minutes, secs, millis);
    }

    public List<SubtitleEntry> parseSrt(Path srtPath) throws Exception {
        String content = Files.readString(srtPath, StandardCharsets.UTF_8);
        String[] blocks = content.trim().split("\n\n");
        List<SubtitleEntry> entries = new ArrayList<>();

        for (String block : blocks) {
            String[] lines = block.trim().split("\n");
            if (lines.length < 3) continue;

            int index = Integer.parseInt(lines[0].trim());
            String timeLine = lines[1].trim();
            String[] times = timeLine.split(" --> ");
            String start = times[0].trim();
            String end = times[1].trim();

            String sourceText = "";
            String translatedText = "";
            if (lines.length >= 4) {
                sourceText = lines[2].trim();
                translatedText = lines[3].trim();
            } else if (lines.length == 3) {
                translatedText = lines[2].trim();
            }

            SubtitleEntry entry = new SubtitleEntry();
            entry.setIndex(index);
            entry.setStartTime(start);
            entry.setEndTime(end);
            entry.setSourceText(sourceText);
            entry.setTranslatedText(translatedText);
            entries.add(entry);
        }

        return entries;
    }
}
